sled-agent/src/instance_manager.rs

```diff
@@ -320,6 +327,27 @@ impl InstanceManager {
             .await
             .map_err(Error::from)
     }
+
+    /// Create a zone bundle from a named instance zone, if it exists.
+    pub async fn create_zone_bundle(
+        &self,
+        name: &str,
+    ) -> Result<ZoneBundleMetadata, BundleError> {
+        for (_propolis_id, instance) in
+            self.inner.instances.lock().await.values()
```

Comment:
Since the Propolis ID is part of the value in the map, I wonder if we could instead do something like this:

```rust
if let Some(instance) =
    self.inner
        .instances
        .lock()
        .unwrap()
        .values()
        .find(|(pid, _)| name == propolis_zone_name(pid))
        .cloned()
{
    match instance.request_zone_bundle(name).await {
        // we already know the name matches
        InstanceZoneBundleResult::NotMe => unreachable!(),

        // other cases as before
    }
}
```

Cloning the `Instance` is OK because it's just a wrapper around an `Arc` that contains the actual `InstanceInner`. The other instance manager routines do basically this to get around the sync/async mismatch; the difference is that they're looking up instances directly by key instead of trying to look for matching values, so they can call `get` whereas this code will have to `find`.

I think this allows you not to hold the instance manager lock while taking the bundle, which (if I'm understanding everything else correctly) should be enough to avoid having to turn that lock into a Tokio mutex. (It also saves us from holding a big lock over a possibly-long-running file copy operation.)

Reply:
That's a great suggestion, thanks Greg. All the contortions were definitely predicated on not cloning, and so needing to hold both locks. That's not great, and thanks to this idea, not needed!

I think this should also get rid of the `InstanceZoneBundleResult`, since the `NotMe` variant will be obviated by the call to `find()`, and we can report `Unavailable` as a `BundleError` directly.

Thanks!

Reply:
So this exact construction doesn't quite work, because the lock on the instance map is still in scope inside the if-let. However! We can use let-else instead:

```rust
        let Some((_propolis_id, instance)) = self
            .inner
            .instances
            .lock()
            .unwrap()
            .values()
            .find(|(propolis_id, _instance)| name == propolis_zone_name(propolis_id))
            .cloned()
        else {
            return Err(BundleError::NoSuchZone { name: name.to_string() });
        };
        instance.request_zone_bundle().await
```

I really do love that syntax addition.